file-watcher-watchdog: add command to kill the operand process on file changes

mfojtik opened the pull request:
The purpose of this package is to live aside to an operand container (side-car). If the operand pod runs with `shareProcessNamespace: true` the watchdog can send SIGTERM to the operand to trigger graceful shutdown.

The watchdog use the fileobserver to detect changes in files (`--files`) and in reaction to a change it will send TERM signal to a process specified by name `--process-name`.

mfojtik commented:
/cc @deads2k 
/cc @sttts 
/cc @tnozicka

sttts commented:
PID namespaces are available as beta since 1.12 🎉

Following the naming discussion on slack: file-change-monitor or file-change-watchdog are better names. Also this is no suicide anymore as the sidecar kills another process.

mfojtik commented:
@sttts @deads2k updated.

I think I want to fully use the shared PID namespace here and use the `/proc/N/root/var/...` trick to check the files. That will reduce the copy&paste for container volume mounts.

I already adjusted the logic here, just need to add the proc path prefix in the sweet spot.

I also added unit tests to prove this works (at least in theory ;-).

mfojtik commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,178 @@
+
+	// Wait while we get the initial PID for the process
+	klog.Infof("Waiting for process %q ...", o.ProcessName)
+	<-pidObservedCh
```
this is the sweet spot... we can prefix all Files with `/proc/getCurrentPID()/root/` :-)

mfojtik replied:
when the PID change, the watchdog will shutdown (and container will restart)... it will get the new PID and prefix the files again.

sttts commented:
I like this, also with the /proc file access.

mfojtik commented:
@sttts implementation finished, i'm going to test this live tomorrow morning with kube apiserver

mfojtik commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+	fs.StringVar(&o.ProcessName, "process-name", "", "name of the process to send signal to on file change (eg. 'hyperkube').")
+	fs.StringSliceVar(&o.Files, "files", o.Files, "comma separated list of file names to monitor for changes")
+	fs.StringVar(&o.KubeConfig, "kubeconfig", o.KubeConfig, "kubeconfig file or empty")
+	fs.StringVar(&o.Namespace, "namespace", o.Namespace, "namespace to report the watchdog events")
```
this is used only for events... as this will restart on every change, I think we **need** events to confirm this works as expected.

mfojtik commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+		// get the monitored process PID
+		observedPID, found, err := o.getPIDByNameFn(o.ProcessName)
+		if !found {
+			klog.Warningf("Process %q does not have PID", o.ProcessName)
```
(this means the watchdog will restart until it gets the process)

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+	ProcessName string
+	Files       []string
+	KubeConfig  string
+	Namespace   string
```
which namespace is this about? godoc

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+
+	// for unit-test
+	getPIDByNameFn       func(name string) (int, bool, error)
+	signalFn             func(pid int) error
```
what is this? godoc

mfojtik replied:
thought the `for unit-test` applies to all below ;-)

mfojtik replied:
i will add them

sttts replied:
it's more like `handleFileChangeFn(pid int)`, right?

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+	fs.StringSliceVar(&o.Files, "files", o.Files, "comma separated list of file names to monitor for changes")
+	fs.StringVar(&o.KubeConfig, "kubeconfig", o.KubeConfig, "kubeconfig file or empty")
+	fs.StringVar(&o.Namespace, "namespace", o.Namespace, "namespace to report the watchdog events")
+	fs.DurationVar(&o.Interval, "interval", 5*time.Second, "interval specify how aggressive the file checks should be")
```
specifying

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+	return nil
+}
+
+// startPidObserver runs a loop that observe changes to the PID of the process we send signals after change is detected.
```
observes

sttts replied:
changes to the PID?

mfojtik replied:
yeah, when the PID change this will shutdown() or send the pid to the control logic

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+
+// startPidObserver runs a loop that observe changes to the PID of the process we send signals after change is detected.
+// When the PID is observed for the first time, the pidObservedCh is closed.
+// When the PID change is observed (after we got initial PID), the shutdown is called (to signal watchdog to restart).
```
why do we restart? Isn't there a chance this way that the watchdog misses file changes?

* operand starts
* sidecar start
* secret changes
* sidecar kills operand
* sidecar terminates, operand terminates
* operand start
* secret changes again
* sidecar start
=> boom. Latest change not noticed.

sttts replied:
tl/dr: I don't like the sidecar to restart at all. It shouldn't.

mfojtik replied:
I'm more worried about the watchdog to be "stucked" or reading the files from wrong directory on PID change. If we don't want to restart, it means we have to make the logic in the `Run()` more complex and less error prone.

Why you hate restarts?

sttts replied:
restarts mean races. I don't like races.

mfojtik replied:
ok, lets go with no restart, I updated the code to run in loop, contexts are fun.

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+
+// readInitialFileContent reads the content of files specified.
+// This is needed by file observer.
+func (o *FileWatcherOptions) readInitialFileContent() error {
```
how do we make sure that this is the state the operand sees? We can be late, and miss an update.

mfojtik replied:
If we restart, yes. Practically I don't see a big chance to miss a change (we rotate certs on hour basis). IF this will be monitoring configs, then yes.

We can discuss restart vs. monitor via Slack.

mfojtik replied:
lets go with the monitor for now (no crashing)

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+// setFilesPathPrefix mutates the file list and prefix every file with /proc/PID/root.
+func setFilesPathPrefix(oldFiles []string, pid int) []string {
```
this is not a setter. Call it `addProcPrefix` or something like that.

mfojtik replied:
sure

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+
+	// Read initial file content. If shared PID namespace does not work, this will error.
+	if err := o.readInitialFileContent(); err != nil {
+		o.recorder.Warningf("FileChangeWatchdogFailed", "Reading initial file content failed: %v", err)
```
does this really need events? It's a misconfiguration. I think logs are enough.

mfojtik replied:
if we restart/crash the log is gone, don't we learned our lessons ? :-)

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+		return fmt.Errorf("unable to read initial file content: %v", err)
+	}
+
+	o.recorder.Eventf("FileChangeWatchdogStarted", "Started watching files for process %s[%d]", o.ProcessName, currentPID)
```
these is are too many events. I only see one event we need: "file xyz changed, restarting operand".

mfojtik replied:
Without aggregated logging, the events are the only way we can discover problems or breakages. If I can get logs from all watchdog containers (even the ones that died), I would remove events.

sttts replied:
Feels like we misuse events as logs. Events should be user relevant. But our events are for us.

mfojtik replied:
Who is the user in this case? I think the target audience for these events are "admins" who are trying to figure out why their operands are not restarting? :-)

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,247 @@
+		// When file change is observed, don't give up on errors when we fail to send signal, but retry sending
+		// and fail hard if we fail to deliver it.
+		retries := 0
+		return wait.PollImmediate(1*time.Second, 5*time.Second, func() (done bool, err error) {
```
what are these intervals? Why those numbers?

mfojtik replied:
I will add godoc (make 5 attempts to SIGINT, wait 1s between them... This might not be needed, but i'm skeptical that things will always work and this is a critical path for watchdog, hence the retry.

sttts commented on pkg/operator/watchdog/proc.go:
```diff
@@ -0,0 +1,52 @@
+// GetPidByName find the process name specified by name and return the PID of that process.
+// If the process is not found, the bool is false.
+// NOTE: This require container with shared process namespace (if run as side-car).
+func GetPidByName(name string) (int, bool, error) {
```
FindPid?

mfojtik replied:
It is by using name... golang already have funcs to get pid, but not by process name... `FindPidByName` ?

sttts replied:
FindPidByName is good

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,295 @@
+// startPidObserver runs a loop that observes changes to the PID of the process we send signals after change is detected.
+// When the PID is observed for the first time, the pidObservedCh is closed.
+// When the PID change is observed (after we got initial PID), the shutdown is called (to signal watchdog to restart).
+func (o *FileWatcherOptions) startPidObserver(ctx context.Context, pidObservedCh chan int) {
```
"start" sounds like this will not block.

mfojtik replied:
will replace with `run`

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,295 @@
+	Interval time.Duration
+
+	// for unit-test to mock getting the process PID
+	handleFindPidByNameFn func(name string) (int, bool, error)
```
no need for the "handle" prefix.

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,295 @@
+	handleSignalFn func(pid int) error
+
+	// for unit-test to mock prefixing files (/proc/PID/root)
+	handleAddProcPrefixToFilesFn func([]string, int) []string
```
no need for the "handle" prefix

mfojtik commented on pkg/operator/watchdog/proc.go:
```diff
@@ -0,0 +1,57 @@
+// This is used to read broken links as in case of share PID namespace, the /proc/1/exe points to a binary
+// that does not exists from the source container.
+func readlink(name string) string {
+	for l := 128; ; l *= 2 {
```
@soltysh will `os.Readlink` return the link target AND the error to ignore? i remember I tried several ways and all of them returned empty string :-(

mfojtik replied:
added unit test to cover all cases we need readlink() for... if it breaks (on any platform) we will know.

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,316 @@
+}
+
+func (o *FileWatcherOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&o.ProcessName, "process-name", "", "name of the process to send signal to on file change (eg. 'hyperkube').")
```
which signal? And only once?

mfojtik replied:
do we want to give user choice? it is TERM

sttts replied:
no, but document :)

SIGTERM is pretty aggressive. I.e. not graceful shutdown. Intentional?

sttts replied:
normally: SIGINT, wait seconds, SIGTERM. A configurable delay would be good (again, graceful shutdown).

sttts commented on pkg/operator/watchdog/proc.go:
```diff
@@ -0,0 +1,57 @@
+
+// readlink is copied from the os.Readlink() but does not return error when the target path does not exists.
+// This is used to read broken links as in case of share PID namespace, the /proc/1/exe points to a binary
+// that does not exists from the source container.
```
omg

mfojtik replied:
https://github.com/golang/go/pull/31994 :-)

soltysh replied:
Yeah, I've had very long discussion with @mfojtik about it, but there's no way, for now. He opened a go issue, iirc.

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,320 @@
+			defer func() {
+				shutdown()
+			}()
+			o.recorder.Eventf("FileChangeObserved", "Observed change in file %q, sending SIGINT to %s[%d]", file, o.ProcessName, currentPID)
```
INT or TERM?

sttts replied:
not that the API server will terminate non-gracefully on the second signal. Does this mean it has only 1 second to terminate gracefully?

mfojtik replied:
I added graceful shutdown sequence

sttts commented on pkg/operator/watchdog/cmd.go:
```diff
@@ -0,0 +1,345 @@
+	observer.AddReactor(func(file string, action fileobserver.ActionType) error {
+		// We already signalled this PID to terminate and the process is being gracefully terminated now.
+		// Do not duplicate termination process for PID we already terminated, but wait for the new PID to appear.
+		if currentPID == o.lastTerminatedPid {
```
add something here that PIDs change on restart.

sttts commented:
/approve
/lgtm

sttts commented:
unit test
